### PR TITLE
Add SOASTA as a builtin analytics vendor

### PR DESCRIPTION
**Pull request by querymetrics**
ITI: https://github.com/ampproject/amphtml/issues/2683

### CODE REVIEW

**Comment by avimehta** [comment]
@querymetrics Thanks for the change! Can you reply here when this is ready for review?

File: src/url-replacements.js [hunk 1]
```diff
@@ -324,8 +352,10 @@ class UrlReplacements {
   /**
    * Returns navigation timing information based on the start and end events.
    * The data for the timing events is retrieved from performance.timing API.
+   * If start and end events are both given, the result is the difference between the two.
+   * If only start event is given, the result is the timing value at start event.
    * @param {string} startEvent
-   * @param {string} endEvent
+   * @param {string} [endEvent]
```
Comment (avimehta) [inline src/url-replacements.js]: I think this repo uses the google closure style docs: http://usejsdoc.org/tags-param.html#optional-parameters-and-default-values

File: extensions/amp-analytics/0.1/vendors.js [hunk 2]
```diff
@@ -39,6 +39,9 @@ export const ANALYTICS_CONFIG = {
       'documentReferrer': 'DOCUMENT_REFERRER',
       'domainLookupTime': 'DOMAIN_LOOKUP_TIME',
       'domInteractiveTime': 'DOM_INTERACTIVE_TIME',
+      'navRedirectCount': 'NAV_REDIRECT_COUNT',
+      'navTiming': 'NAV_TIMING',
```
Comment (avimehta) [inline extensions/amp-analytics/0.1/vendors.js]: This is a much better and flexible way of doing this. Thanks for adding this!

File: extensions/amp-analytics/amp-analytics.md [hunk 3]
```diff
@@ -158,6 +158,18 @@ Type attribute value: `quantcast`
...
+### SOASTA mPulse
+
+Type attribute value: `mpulse`
+
+Adds support for [SOASTA mPulse](https://www.soasta.com/mPulse). Configuration details can be found at [docs.soasta.com](http://docs.soasta.com/).
+Vars `ab_test` and `site_version` are optional.
+
+Example of usage with license key of `XXXXX-XXXXX-XXXXX-XXXXX-XXXXX`:
+```html
+<amp-analytics type="mpulse" config="https://c.go-mpulse.net/concerto/api/amp-config.json?d=SOURCE_HOST&key=XXXXX-XXXXX-XXXXX-XXXXX-XXXXX"></amp-analytics>
```
Comment (avimehta) [inline extensions/amp-analytics/amp-analytics.md]: the & is shown in red. Does replacing it with `&amp;` fix it?

File: extensions/amp-analytics/analytics-vars.md [hunk 4]
```diff
@@ -206,6 +206,46 @@ is unloaded. The value is in milliseconds.
...
+Please see below for the required and optional arguments you may pass into `navTiming` like a function. Spaces between arguments and values are not allowed.
+
+**arguments**:
+
+  - `start-event` (Required) - Name of the PerformanceTiming interface attribute corresponding to the start event.
+
+  - `end-event` (Optional) - Optional name of the PerformanceTiming interface attribute corresponding to the end event. If `end-event` is passed, the value will be the time difference between the start and end events.
+
+
+Example 1 usage: `${navTiming(navigationStart)}`
```
Comment (avimehta) [inline extensions/amp-analytics/analytics-vars.md]: nit: Maybe reword this as : 

Example 1: `${navTiming(navigationStart)}` results in value: `1451606400000`

Example 2:...

File: src/url-replacements.js [hunk 5]
```diff
@@ -301,6 +301,34 @@ class UrlReplacements {
...
+      if (!navigationInfo || navigationInfo['type'] === undefined) {
+        // PerformanceNavigation interface is not supported.
+        return Promise.resolve();
+      }
+
+      return String(navigationInfo['type']);
+    });
+
+    this.set_('NAV_REDIRECT_COUNT', () => {
+      const navigationInfo = this.win_['performance']
```
Comment (avimehta) [inline src/url-replacements.js]: can this method and the one above share the code?

**Comment by querymetrics** [comment]
@avimehta thank you for the feedback! I made some changes based on your comments.

**Comment by querymetrics** [comment]
@avimehta, it's ready for review, thanks!

**Comment by nicjansma** [comment]
@avimehta thanks for your help in reviewing this.  Do you have any immediate concerns we should look into?

File: extensions/amp-analytics/amp-analytics.md [hunk 6]
```diff
@@ -164,6 +164,64 @@ Type attribute value: `quantcast`
 
 Adds support for Quantcast Measurement. More details for adding Quantcast Measurement can be found at [quantcast.com](https://www.quantcast.com/help/guides/)
 
+### SOASTA mPulse
+
+Type attribute value: `mpulse`
+
+Adds support for [SOASTA mPulse](https://www.soasta.com/mPulse). Configuration details can be found at [docs.soasta.com](http://docs.soasta.com/).
+
+Optional `extraUrlParams` parameters include `page_group`, `ab_test`, `custom_dimension.*` and `custom_metric.*` which are defined below. Note that since these optional parameters are included in the page output they may be cached by AMP proxy caches and served to multiple visitors.
```
Comment (rudygalfi) [inline extensions/amp-analytics/amp-analytics.md]: I'm concerned that adding this level of documentation detail here makes this page harder to parse. Would it be possible to fold this into documentation that lives at docs.soasta.com?

**Comment by avimehta** [comment]
@nicjansma Sorry for the delay. The PR looks good now (one minor comment). Can you fix the names and then squash your commits into one? Also rebase to head?

File: extensions/amp-analytics/analytics-vars.md [hunk 7]
```diff
@@ -206,6 +206,42 @@ is unloaded. The value is in milliseconds.
...
+### navRedirectCount
+
+Provides the number of redirects since the last non-redirect navigation.
+See W3C Navigation Timing API [PerformanceNavigation interface](https://www.w3.org/TR/navigation-timing/#performancenavigation) for more information.
+
+Example value: `0`
+
+### navTiming
+
+Provides access to metrics from the browser's PerformanceTiming interface.
```
Comment (avimehta) [inline extensions/amp-analytics/analytics-vars.md]: Can you change start-event to startEvent and end-event to endEvent? that way, we'll have consistency in naming. Here and below.

**Comment by querymetrics** [comment]
thanks @avimehta, squashed and rebased.

**Comment by avimehta** [comment]
@dvoytenko @jridgewell The build seems to have failed for a weird reason. Can you kick it again?

**Comment by jridgewell** [comment]
Kicked.

**Comment by nicjansma** [comment]
@avimehta With this reviewed and all checks passed, what would be the timeline for getting this integrated?

**Comment by avimehta** [comment]
@nicjansma There is nothing left from the process point of view. This PR is ready to merge. I have been having issue with merging PRs recently though. 

@dvoytenko/@cramforce: Can you merge this PR?

**Comment by nicjansma** [comment]
Great!  Thanks all.

File: extensions/amp-analytics/0.1/vendors.js [hunk 8]
```diff
@@ -541,6 +544,73 @@ export const ANALYTICS_CONFIG = {
       'image': true,
     },
   },
+
+  'mpulse': {
+    'requests': {
+      'onvisible': 'https://${beacon_url}?' +
```
Comment (lannka) [inline extensions/amp-analytics/0.1/vendors.js]: @querymetrics @avimehta what would this var `beacon_url` resolve to? I'm seeing exception for `mpulse` in the `examples/analytics-vendors.amp.html`

Comment (avimehta) [inline extensions/amp-analytics/0.1/vendors.js]: I thought this was something that is specific to each publisher and hence not specified here.

Comment (lannka) [inline extensions/amp-analytics/0.1/vendors.js]: Could be. So we need to fix the example.
